### ISSUE TITLE
torch.save does not save the classifier

### ISSUE THREAD

**Issue by chenlingantelope** [YosefLab/scVI#225]
I was running SCANVI using the PBMC12K and PBMC68K data and I trained the SCANVI model using the following code:
`vae = VAE(all_dataset.nb_genes, n_batch=all_dataset.n_batches, n_labels=all_dataset.n_labels,
          n_hidden=128, n_latent=10, n_layers=2, dispersion='gene')

trainer = UnsupervisedTrainer(vae, all_dataset, train_size=1.0)
trainer.train(n_epochs=100)
scanvi = SCANVI(all_dataset.nb_genes, all_dataset.n_batches, all_dataset.n_labels, n_layers=2)
scanvi.load_state_dict(trainer.model.state_dict(), strict=False)
trainer_scanvi = SemiSupervisedTrainer(scanvi, all_dataset, classification_ratio=50,
                                       n_epochs_classifier=1, lr_classification=5 * 1e-3)
trainer_scanvi.train(n_epochs=20)
`

I get prediction accuracy of over 0.98. However when I save the model using 
`torch.save(trainer_scanvi.model,'scanvi.model.pkl')`
and instead of training the second time I load this experiment for analysis, I load the existing model: 
`trainer_scanvi.model = torch.load('scanvi.model.pkl')`
I get about 0.06 label prediction accuracy. so it seems like torch.save didn't save all of the parameters although I am not sure which ones was missing. Another possibility is somehow the cluster identity got scrambled somewhere because after reloading the model the scanvi latent space still looked clustered by cell type except that the cell type proportions and positions are very off. 
![scanvi_scrambled](https://user-images.githubusercontent.com/4098673/49109524-cbbfb480-f23f-11e8-8719-fb8d7cfa600b.png)
The colors for PBMC should look more like this following plot: 
[Easy1.scanvi1.labels.pdf](https://github.com/YosefLab/scVI/files/2621733/Easy1.scanvi1.labels.pdf)

**Comment by romain-lopez** [comment]
When I try to get the parameters list from SCANVI, I get some redundancy:

`(classifier): Classifier(
    (layers): FCLayers(
      (fc_layers): Sequential(
        (Layer 0): Sequential(
          (0): Linear(in_features=10, out_features=128, bias=True)
          (1): BatchNorm1d(128, eps=0.001, momentum=0.01, affine=True, track_running_stats=True)
          (2): ReLU()
          (3): Dropout(p=0.1)
        )
        (Layer 1): Sequential(
          (0): Linear(in_features=128, out_features=128, bias=True)
          (1): BatchNorm1d(128, eps=0.001, momentum=0.01, affine=True, track_running_stats=True)
          (2): ReLU()
          (3): Dropout(p=0.1)
        )
      )
    )
    (classifier): Sequential(
      (0): FCLayers(
        (fc_layers): Sequential(
          (Layer 0): Sequential(
            (0): Linear(in_features=10, out_features=128, bias=True)
            (1): BatchNorm1d(128, eps=0.001, momentum=0.01, affine=True, track_running_stats=True)
            (2): ReLU()
            (3): Dropout(p=0.1)
          )
          (Layer 1): Sequential(
            (0): Linear(in_features=128, out_features=128, bias=True)
            (1): BatchNorm1d(128, eps=0.001, momentum=0.01, affine=True, track_running_stats=True)
            (2): ReLU()
            (3): Dropout(p=0.1)
          )
        )
      )
      (1): Linear(in_features=128, out_features=9, bias=True)
      (2): Softmax()
    )
  )`

While t

**Comment by romain-lopez** [comment]
That is not readable but there are two classifiers here, I think it comes from the tracking system of PyTorch that gets this twice in the classifier's constructor:

`        self.layers = FCLayers(n_in=n_input, n_out=n_hidden, n_layers=n_layers, n_hidden=n_hidden,
                               dropout_rate=dropout_rate, use_batch_norm=True)
        self.classifier = nn.Sequential(self.layers, nn.Linear(n_hidden, n_labels), nn.Softmax(dim=-1))`

Maybe concatenating these lines would already avoid saving twice the same parameters ?

**Comment by romain-lopez** [comment]
I'm going to run some tests but I think the saving pipeline is fine besides the glitch I found. It might be you do not call model.eval() ?

`# Model class must be defined somewhere
model = torch.load(PATH)
model.eval()`

**Comment by romain-lopez** [comment]
Yeah it looks like the loading module is working well. @chenlingantelope  would you try again following the instructions from https://pytorch.org/tutorials/beginner/saving_loading_models.html#saving-loading-model-for-inference and let me know ?

**Comment by chenlingantelope** [comment]
Thanks @romain-lopez I just tried using these commands instead and the model loads correctly now! 
`trainer_scanvi.train(n_epochs=10)
torch.save(trainer_scanvi.model.state_dict(),'scanvi.model.pkl')
trainer_scanvi.model.load_state_dict(torch.load('scanvi.model.pkl'))
trainer_scanvi.model.eval()`
